### PR TITLE
Parallel reloading of ingress and egress eBPF programs

### PR DESCRIPTION
**Pull request by alexkats**
**Background**
One of the Kubernetes SLO is the Pod Startup Latency SLO. The
current limit is the 99th percentile <= 5s. With the use of cilium
Kubernetes is very close to this limit, sometimes even exceeding it.
Reloading of ingress and egress eBPF programs in parallel is one of the
efforts towards reducing this Pod Startup Latency.

**Test scenario**
All tests were run on the cluster version 1.23.7-gke.500. The 1.11 cilium version was used as a baseline for all the tests.

To test different configurations a GKE sandbox with custom Dataplane v2 images was used. The cluster was created with 100 nodes and DPv2 enabled. For Pod Startup Latency measurements a modified configuration of a GKE Scalability 100 nodes performance test was used.

In the beginning for each run 1 pod was created on each node with the same image as in Pod Startup Latency measurement, so image pull time is not affecting the resulting latency. Then 500 deployments with 5 pods each were created and waited until running to measure the pod startup latency.

**Results**
Percentile | Baseline | Parallel reloading
--- | --- | ---
P50 | 3162 ms | 3001 ms
P90 | 3745 ms | 3576 ms
P99 | 4121 ms | 3968 ms

**Summary**
Parallel reloading optimization may seem not a big deal, but it still reduces the total latency by ~150ms. It also gives some room in case reloading of ingress and egress eBPF programs becomes more time consuming.

Signed-off-by: Alex Katsman <alexkats@google.com>

### CODE REVIEW

**Comment by joestringer** [comment]
Thanks for the proposal! I wonder if you were able to use the `cilium_endpoint_regeneration_time_stats_seconds` metric to gather more understanding about how Cilium is spending that time during regeneration, in hopes that we can further optimize?

File: pkg/datapath/loader/loader.go
Comment (alan-kut) [inline pkg/datapath/loader/loader.go]: Can it be extracted into separate function using go routine and used for both ingress and egress?

Comment (Weil0ng) [reply pkg/datapath/loader/loader.go]: +1

Comment (alexkats) [reply pkg/datapath/loader/loader.go]: Updated

**Comment by alexkats** [comment]
I didn't use this metric, thanks, I'll take a look into it. There are more optimizations coming which are going to reduce eBPF program verification time for Kubernetes. This PR is the first step towards reducing pod startup latency.

The kernel version that was used is 5.10.109.

**Comment by alexkats** [comment]
Ok, I see, makes sense. Thanks for the info.

File: pkg/datapath/loader/loader.go
Comment (alan-kut) [inline pkg/datapath/loader/loader.go]: I have a mixed feelings here. I'd wrap it into go rountine here instead of just pushing results to chans.

Comment (alexkats) [reply pkg/datapath/loader/loader.go]: Wrapped it in a goroutine here

**Comment by alexkats** [comment]
Hi @ti-mo, could you please take a look at this proposal and/or reroute it to someone else for review.

**Comment by ti-mo** [comment]
Hi @alexkats, sorry for the delay here. Had some PTO followed by illness.

Honestly, I'm a little torn about this change and have been mulling it over for a few days. This code is already quite complex and hard to grok, so my plan was to reduce the amount of async stuff we do around these packages, not increase it.

I'm also questioning the results of your benchmark. You state 1.11 as the baseline for your comparison, but if you're comparing to master, that's a huge jump with too many factors to account for. One notable PR that was merged recently is #19159, replacing ip/tc shellouts with native Go code. In my measurements, the Go ELF loader is faster than the version of libbpf used by iproute2, which also skews the results. There have also been a few tailcall splits merged into 1.13 that drive down verification time.

As Joe already pointed out, the kernel has a global verifier lock, with program verification being responsible for a large chunk of endpoint creation time (assuming a build cache hit). Throwing concurrency at the problem won't improve this. tc operations are on the order of microseconds under normal circumstances, so unlikely to matter. Something on the list of potential improvements is removing ELF rewriting, meaning a built ELF only needs to be parsed from disk into a CollectionSpec once, and can be rewritten in-memory from then on just before loading into the kernel.

Side note: assuming this pod is the first endpoint of its kind to land on a node, those 4s likely include the compile time of the bpf prog used for the container interface. I don't think this is representative of Cilium's endpoint creation performance at large.

Since all loader code is now Go-based, pprof should give a clear picture of where time is actually spent, making optimization hunting much easier. If using pprof is not an option, I'd be interested in seeing the difference against master as a baseline. A 3.7% improvement is nice, but barely outside margin of error and not sure if worth the additional complexity.

**Comment by alexkats** [comment]
Hi @ti-mo, thank you so much for your reply.

So, I re-ran tests from master with cluster version 1.24.4-gke.300 (I was unable to run it successfully beforehand, that's why all tests were ran on 1.11). Here's a table with updated results:

Percentile | Baseline | Parallel reloading
--- | --- | ---
P50 | 2980 ms | 2830 ms
P90 | 3569 ms | 3409 ms
P99 | 3938 ms | 3781 ms

So master version is better than 1.11 by ~180ms in terms of Pod Startup Latency. But parallelization still reduces it by ~150ms more, which looks significant as we aim to go below 3.5 seconds eventually for P99.

I'd like also to elaborate a bit on tests. Actually there were some pods created on each node before starting measurements in order to exclude other factors, such as image pull time or bpf program compilation. And from logs I see that there was no compilation happening during these 4 seconds. Also regarding margin of error, I ran tests multiple times (20-25), and presented results are average of all runs, so to me it seems to be pretty stable improvement.

Regarding adding more complexity to the code, I see your point, but to me it doesn't seem that much, as it's a moving kinda duplicated code into a function with adding goroutine. I could also try to use pprof for further investigation in order to optimize loader part more, but anyway it seems that this change could be a good start.

What are your thoughts on this?

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [inline pkg/datapath/loader/loader.go]: Could https://pkg.go.dev/golang.org/x/sync/errgroup simplify things here? `asyncReplaceDatapath` could take a `chan func() error` it pushes finalizers onto that can be drained in a loop. That would result in a single error check here and tidies things up a bit.

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [inline pkg/datapath/loader/loader.go]:
```suggestion
				scopedLog.WithError(err).Warnf("JoinEP: Failed to attach %s program", progDirection)
```

**Review by ti-mo — CHANGES_REQUESTED** [review]
Thanks again for pushing this. Clearly this optimization holds merit, but before we commit to this, I'd still like to know exactly where this 150ms time saving comes from.

I'm saying this as this might influence how we architect improvements to the loader pipeline going forward. It might make sense to incorporate more async/'batch' APIs at a lower level instead of burdening the business logic at this level of abstraction with spawning goroutines etc. Conversely, if we notice this 150ms is spent parsing the ELF, this particular optimization might be redundant and we should focus on minimizing ELF loads by passing around `CollectionSpec`s instead.

@alexkats is this something you could figure out somehow?

> we aim to go below 3.5 seconds eventually for P99

This sounds rather arbitrary, what's with 3.5s?

**Comment by alan-kut** [comment]
> This sounds rather arbitrary, what's with 3.5s?

We want to minimize the pod startup latency regression when changing kube-proxy based dataplane to Cilium on GKE. \
There is a k8s SLO - 99th percentile per cluster-day <= 5s, we have room to spend but we want to use as little as possible of it.

The code is simplified here, except for the additional go routine but at the end I don't see any downside of this go routine. If parsing ELF can be optimized then when it is those go routines can be removed.

File: pkg/datapath/loader/loader.go
Comment (alexkats) [reply pkg/datapath/loader/loader.go]: Thanks for pointing out, used errgroup here

File: pkg/datapath/loader/loader.go
Comment (ti-mo) [inline pkg/datapath/loader/loader.go:298]:
```suggestion
			return err
		}
		finalizeChan <- finalize
		return nil
```

If err is non-nil, `finalize` will be nil and the range over `finalizers` will try to invoke it.

**Review by ti-mo — CHANGES_REQUESTED** [review]
Thanks, one thing left to address. Will kick off a round of tests in the meantime.

Edit: feel free to squash so that doesn't need to be done later.

File: pkg/datapath/loader/loader.go
Comment (alan-kut) [reply pkg/datapath/loader/loader.go:298]: I think that it won't be the case as the function will exit with just an error, without invoking any finalizers if there is any error here:

```
if err := wg.Wait(); err != nil {
	return err  // <--- HERE
}

close(finalizers)
for finalize := range finalizers {
	defer finalize()
}
```

Still I agree with the proposed change as it's cleaner.

File: pkg/datapath/loader/loader.go
Comment (alan-kut) [inline pkg/datapath/loader/loader.go]: In the original code if one program replaces successfully it's finalizer will be called and the function will proceed to the second one.
After the change both of them are process in parallel.

Should we still defer finalizer for replacements that finished successfully or only if all replacement succeeded?

File: pkg/datapath/loader/loader.go
Comment (alexkats) [reply pkg/datapath/loader/loader.go:298]: Applied

File: pkg/datapath/loader/loader.go
Comment (alexkats) [reply pkg/datapath/loader/loader.go]: It seems fine to me, but I can easily change if we need to finalize successful replacement. @ti-mo could you comment on this one, please?

Comment (ti-mo) [reply pkg/datapath/loader/loader.go]: @alan-kut Good catch.

> if one program replaces successfully it's finalizer will be called

If the first prog attaches successfully, it will be deferred, not immediately called.

> Should we still defer finalizer for replacements that finished successfully or only if all replacement succeeded?

Yes, the first replacement might return a finalizer and the second one could fail. The finalizer of the first one still needs to be run. The correct ordering here should be:

```
err := wg.Wait()
<close & range finalizers>
if err != nil {
  return err
}
```

**Review by alan-kut — APPROVED** [review]

**Review by ti-mo — APPROVED** [review]
Thanks, this looks good. Will wait for CI to go green before marking ready-to-merge.

**Comment by ti-mo** [comment]
As I expected, loading both CollectionSpecs concurrently races to pin the tail call map for the endpoint. Initially, I thought this would break map migration on up/downgrade, but it seems we ignore missing map pins on removal.

Unfortunately, there's no straightforward fix for this aside from doing some of the refactoring I mentioned above. Both of these ingress/egress progs should be attached from the same Collection, meaning `replaceDatapath` should be changed to take a `Collection` instead of an `objPath`. If the Collection is loaded only once, this parallel attachment also becomes unnecessary, since the progs will only be verified once, which is the lion's share of the time gain here. Right now, the same ELF is parsed and loaded into the kernel (and verified) twice, one for each program.

**Comment by alexkats** [comment]
I see the problem, thanks for the detailed explanation of what happened and your help overall. I'm going to close this PR for now, and focus on other improvements.

**Comment by ti-mo** [comment]
@alexkats Thanks for understanding, sorry this didn't work out. This exercise did bring to our attention that parallel verification of progs is possible, iirc the global verifier lock was removed for privileged users in 5.x kernels. We'll keep this in mind in future improvements!

**Pull request closed without merging**
